File: src/common/py/api/connector/connector_commands.py
import dataclasses
import typing

from ...core.messaging import (
    Command,
    CommandReply,
    Message,
)
from ...core.messaging.composers import (
    MessageBuilder,
    CommandComposer,
    CommandReplyComposer,
)
from ...data.entities import (
    Connector,
)


@Message.define("command/connector/list")
class ListConnectorsCommand(Command):
    """
    Command to fetch all available connectors.

    Notes:
        Requires a ``ListConnectorsReply`` reply.
    """

    @staticmethod
    def build(
        message_builder: MessageBuilder, *, chain: Message | None = None
    ) -> CommandComposer:
        """
        Helper function to easily build this message.
        """
        return message_builder.build_command(ListConnectorsCommand, chain)


@Message.define("command/connector/list/reply")
class ListConnectorsReply(CommandReply):
    """
    Reply to ``ListConnectorsCommand``.

    Args:
        connectors: List of all connectors.
    """

    connectors: typing.List[Connector] = dataclasses.field(default_factory=list)

    @staticmethod
    def build(
        message_builder: MessageBuilder,
        cmd: ListConnectorsCommand,
        *,
        connectors: typing.List[Connector],
        success: bool = True,
        message: str = "",
    ) -> CommandReplyComposer:
        """
        Helper function to easily build this message.
        """
        return message_builder.build_command_reply(
            ListConnectorsReply, cmd, success, message, connectors=connectors
        )

File: src/common/web/api/connector/ConnectorCommands.ts
import { Type } from "class-transformer";

import { Command } from "../../core/messaging/Command";
import { CommandReply } from "../../core/messaging/CommandReply";
import { CommandComposer } from "../../core/messaging/composers/CommandComposer";
import { CommandReplyComposer } from "../../core/messaging/composers/CommandReplyComposer";
import { MessageBuilder } from "../../core/messaging/composers/MessageBuilder";
import { Message } from "../../core/messaging/Message";
import { Connector } from "../../data/entities/connector/Connector";

/**
 * Command to fetch all available connectors.
 */
@Message.define("command/connector/list")
export class ListConnectorsCommand extends Command {
    /**
     * Helper function to easily build this message.
     */
    public static build(messageBuilder: MessageBuilder, chain: Message | null = null): CommandComposer<ListConnectorsCommand> {
        return messageBuilder.buildCommand(ListConnectorsCommand, {}, chain);
    }
}

/**
 * Reply to ``ListConnectorsCommand``.
 *
 * @param connectors - List of all connectors.
 */
@Message.define("command/connector/list/reply")
export class ListConnectorsReply extends CommandReply {
    // @ts-ignore
    @Type(() => Connector)
    public readonly connectors: Connector[] = [];

    /**
     * Helper function to easily build this message.
     */
    public static build(
        messageBuilder: MessageBuilder,
        cmd: ListConnectorsCommand,
        connectors: Connector[],
        success: boolean = true,
        message: string = ""
    ): CommandReplyComposer<ListConnectorsReply> {
        return messageBuilder.buildCommandReply(ListConnectorsReply, cmd, success, message, { connectors: connectors });
    }
}
